src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataReaderExtensions.cs

```diff
@@ -71,7 +72,45 @@ public static ODataItemBase ReadResourceOrResourceSet(this ODataReader reader)
                             "The resource which is ending should be on the top of the items stack.");
                         itemsStack.Pop();
                         break;
+                    case ODataReaderState.DeletedResourceStart:
+                        ODataDeletedResource deletedResource = (ODataDeletedResource)reader.Item;
+                        ODataResourceWrapper deletedResourceWrapper = null;
+                        if (deletedResource != null)
+                        {
+                            deletedResourceWrapper = new ODataResourceWrapper(deletedResource);
+                        }
 
+                        if (itemsStack.Count == 0)
+                        {
+                            Contract.Assert(deletedResource != null, "The top-level resource can never be null.");
+                            topLevelItem = deletedResourceWrapper;
+                        }
+                        else
+                        {
+                            ODataItemBase parentItem = itemsStack.Peek();
```

Comment:
top-level resource should never be deleted.  you should be able to replace these 7 lines with a Contract.Assert(itemsStack.count != 0) and then continue on with line 90.